File: app/src/main/java/name/monwf/customiuizer/utils/PreferenceAdapter.java
package name.monwf.customiuizer.utils;

import java.util.ArrayList;
import java.util.Arrays;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.drawable.Drawable;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import name.monwf.customiuizer.R;

public class PreferenceAdapter extends BaseAdapter {
	private final LayoutInflater mInflater;
	private final String key;
	private final boolean activities;
	private final ArrayList<String> items = new ArrayList<String>();

	public PreferenceAdapter(Context context, String prefKey, boolean actOnly) {
		mInflater = LayoutInflater.from(context);
		key = prefKey;
		activities = actOnly;
		updateItems();
	}

	public void updateItems() {
		items.clear();
		String itemStr = AppHelper.getStringOfAppPrefs(key, "");
		if (itemStr == null || itemStr.isEmpty()) return;
		String[] itemArr = itemStr.trim().split("\\|");
		items.addAll(Arrays.asList(itemArr));
	}

	public int getCount() {
		return items.size();
	}

	public String getItem(int position) {
		return items.get(position);
	}

	public long getItemId(int position) {
		return position;
	}

	@SuppressLint("ClickableViewAccessibility")
	public View getView(final int position, View convertView, ViewGroup parent) {
		View row;
		if (convertView != null)
			row = convertView;
		else
			row = mInflater.inflate(R.layout.pref_item, parent, false);

		Helpers.setMiuiPrefItem(row);
		ImageView dragHandle = row.findViewById(R.id.drag_handle);
		ImageView itemIcon = row.findViewById(android.R.id.icon);
		TextView itemTitle = row.findViewById(android.R.id.title);
		TextView itemSummary = row.findViewById(android.R.id.summary);

		dragHandle.setVisibility(activities ? View.GONE : View.VISIBLE);
		dragHandle.setOnTouchListener(((SortableListView)parent).getListenerForStartingSort());
		itemIcon.setVisibility(View.VISIBLE);
		String uuid = getItem(position);
		Pair<String, String> name = AppHelper.getActionNameLocal(row.getContext(), key + "_" + uuid);
		if (name == null) {
			itemTitle.setText(R.string.notselected);
			itemSummary.setVisibility(View.GONE);
		} else {
			if (activities) {
				String actStr = AppHelper.getStringOfAppPrefs(key + "_" + uuid + "_activity", "");
				itemTitle.setText(actStr == null ? "" : Helpers.getAppName(row.getContext(), actStr, true));
			} else {
				itemTitle.setText(name.first);
			}
			if (name.second == null || name.second.isEmpty()) {
				itemSummary.setVisibility(View.GONE);
			} else {
				itemSummary.setVisibility(View.VISIBLE);
				itemSummary.setText(name.second);
			}
		}

		try {
			Drawable drawable = Helpers.getActionImageLocal(row.getContext(), key + "_" + uuid);
			itemIcon.setImageDrawable(drawable != null ? drawable : row.getContext().getPackageManager().getApplicationIcon(Helpers.modulePkg));
		} catch (Throwable t) {
			t.printStackTrace();
		}

		row.setPadding(row.getPaddingLeft(), row.getPaddingTop(), 0, row.getPaddingBottom());
		return row;
	}

}

File: app/src/main/java/tv/withaibuild/customiuizer/utils/PreferenceAdapter.kt
package tv.withaibuild.customiuizer.utils

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.TextView
import tv.withaibuild.customiuizer.R

class PreferenceAdapter(context: Context, private val key: String, private val activities: Boolean) : BaseAdapter() {

    private val inflater: LayoutInflater = LayoutInflater.from(context)
    private val items = ArrayList<String>()

    init {
        updateItems()
    }

    fun updateItems() {
        items.clear()
        val itemStr = AppHelper.getStringOfAppPrefs(key, "") ?: ""
        if (itemStr.isEmpty()) return
        items.addAll(itemStr.trim().split("|"))
    }

    override fun getCount(): Int = items.size

    override fun getItem(position: Int): String = items[position]

    override fun getItemId(position: Int): Long = position.toLong()

    @SuppressLint("ClickableViewAccessibility")
    override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
        val row = convertView ?: inflater.inflate(R.layout.pref_item, parent, false)

        Helpers.setMiuiPrefItem(row)
        val dragHandle = row.findViewById<ImageView>(R.id.drag_handle)
        val itemIcon = row.findViewById<ImageView>(android.R.id.icon)
        val itemTitle = row.findViewById<TextView>(android.R.id.title)
        val itemSummary = row.findViewById<TextView>(android.R.id.summary)

        dragHandle.visibility = if (activities) View.GONE else View.VISIBLE
        (parent as? SortableListView)?.let { dragHandle.setOnTouchListener(it.getListenerForStartingSort()) }
        itemIcon.visibility = View.VISIBLE
        val uuid = getItem(position)
        val name = AppHelper.getActionNameLocal(row.context, key + "_" + uuid)
        if (name == null) {
            itemTitle.setText(R.string.notselected)
            itemSummary.visibility = View.GONE
        } else {
            if (activities) {
                val actStr = AppHelper.getStringOfAppPrefs(key + "_" + uuid + "_activity", "") ?: ""
                itemTitle.text = Helpers.getAppName(row.context, actStr, true) ?: ""
            } else {
                itemTitle.text = name.first
            }
            if (name.second.isNullOrEmpty()) {
                itemSummary.visibility = View.GONE
            } else {
                itemSummary.visibility = View.VISIBLE
                itemSummary.text = name.second
            }
        }

        try {
            val drawable = Helpers.getActionImageLocal(row.context, key + "_" + uuid)
            itemIcon.setImageDrawable(drawable ?: row.context.packageManager.getApplicationIcon(Helpers.modulePkg))
        } catch (t: Throwable) {
            t.printStackTrace()
        }

        row.setPadding(row.paddingLeft, row.paddingTop, 0, row.paddingBottom)
        return row
    }
}
